.github/workflows/release.yml

```diff
@@ -18,18 +18,67 @@ jobs:
       python_version: ${{ steps.generate-matrix.outputs.python_version }}
       build: ${{ steps.generate-matrix.outputs.build}}
       test: ${{ steps.generate-matrix.outputs.test}}
+      do_macos: ${{ steps.generate-matrix.outputs.do_macos}}
+      do_ubuntu: ${{ steps.generate-matrix.outputs.do_ubuntu}}
+      do_windows: ${{ steps.generate-matrix.outputs.do_windows}}
+      build_doc: ${{ steps.generate-matrix.outputs.build_doc}}
+      deploy_test_pypi: ${{ steps.generate-matrix.outputs.deploy_test_pypi}}
     steps:
       - uses: actions/setup-python@v2
       - name: Generate Matrix
         id: generate-matrix
         shell: python3 {0}
         run: |
+          # Default variables
           python_version = ["3.7", "3.8", "3.9"]
-          build = { "macos": ["macos-10.15"], "linux": ["ubuntu-latest"], "windows": ["windows-2016"] }
-          test = { "macos": ["macos-10.15", "macos-11"], "linux": ["ubuntu-18.04", "ubuntu-20.04"], "windows": ["windows-2016", "windows-2019", "windows-2022"] }
+          build = [ "macos-10.15", "ubuntu-18.04", "windows-2016" ]
+          test = [ "macos-10.15", "macos-11", "ubuntu-18.04", "ubuntu-20.04", "windows-2016", "windows-2019" ]
+          build_doc = "true"
+          deploy_test_pypi = "true"
+
+          to_bool = lambda s: True if s == "true" else False
+          python_filter = {
+              '3.7' : to_bool("${{ contains(github.event.head_commit.message, '[ci: python-3.7]') }}"),
+              '3.8' : to_bool("${{ contains(github.event.head_commit.message, '[ci: python-3.8]') }}"),
+              '3.9' : to_bool("${{ contains(github.event.head_commit.message, '[ci: python-3.9]') }}"),
+          }
+          if any(python_filter.values()):
+              python_version = [v for v in python_version if python_filter[v]]
+          os_filter = {
+              'macos-10.15'  : to_bool("${{ contains(github.event.head_commit.message, '[ci: macos-10.15]') }}"),
+              'macos-11'     : to_bool("${{ contains(github.event.head_commit.message, '[ci: macos-11]') }}"),
+              'ubuntu-18.04' : to_bool("${{ contains(github.event.head_commit.message, '[ci: ubuntu-18.04]') }}"),
+              'ubuntu-20.04' : to_bool("${{ contains(github.event.head_commit.message, '[ci: ubuntu-20.04]') }}"),
+              'windows-2016' : to_bool("${{ contains(github.event.head_commit.message, '[ci: windows-2016]') }}"),
+              'windows-2019' : to_bool("${{ contains(github.event.head_commit.message, '[ci: windows-2019]') }}"),
+          }
+          if set(os_filter.keys()) != set(test):
```

Comment:
Your description states that there is only one keyword parsed by release.yml.  This is obviously false.

Reply:
The statement means that only one keyword is active in `release`. Ideally the setup could/should be the same for both workflow.

Reply:
> The statement means that only one keyword is active in release.

Which is false.  It would be much more obvious if you did not copy everything but only keep the keyword you want to parse.

Reply:
Okay I stand corrected, my statement is wrong, all keywards are active, but needs.setup is never used. But this is error prone, dead code should be removed.